pkg/vault/contracts/CompositeLiquidityRouter.sol

```diff
@@ -482,10 +484,50 @@ contract CompositeLiquidityRouter is ICompositeLiquidityRouter, BatchRouterCommo
         // Loads a Set with all amounts to be inserted in the nested pools, so we don't need to iterate in the tokens
         // array to find the child pool amounts to insert.
         for (uint256 i = 0; i < tokensIn.length; ++i) {
+            if (params.maxAmountsIn[i] == 0) {
+                continue;
+            }
+
             _currentSwapTokenInAmounts().tSet(tokensIn[i], params.maxAmountsIn[i]);
+            _currentSwapTokensIn().add(tokensIn[i]);
         }
 
-        IERC20[] memory parentPoolTokens = _vault.getPoolTokens(params.pool);
+        (uint256[] memory amountsIn, ) = _addLiquidityRecursive(params.pool, params);
+
+        // Adds liquidity to the parent pool, mints parentPool's BPT to the sender and checks the minimum BPT out.
+        (, exactBptAmountOut, ) = _vault.addLiquidity(
+            AddLiquidityParams({
+                pool: params.pool,
+                to: isStaticCall ? address(this) : params.sender,
+                maxAmountsIn: amountsIn,
+                minBptAmountOut: params.minBptAmountOut,
+                kind: params.kind,
+                userData: params.userData
+            })
+        );
+
+        // Settle the amounts in.
+        if (isStaticCall == false) {
+            _settlePaths(params.sender, params.wethIsEth);
+        }
+    }
+
+    function _addLiquidityRecursive(
```

Comment:
```suggestion
    function _addLiquidity(
```
Rather than overloading _addLiquidityRecursive, consider naming the "outer" one _addLiquidity, leaving the "inner" one (that is actually recursive, with the level argument) as _addLiquidityRecursive.